lib/core/viewer/components/img-viewer.component.spec.ts

```diff
@@ -344,11 +148,147 @@ describe('Test Img viewer component ', () => {
+
+            component.zoomOut();
+            expect(component.scale).toBe(0.6);
+            expect(component.cropper.zoom).toHaveBeenCalledWith(-0.2);
+        }));
+
+        it('should not zoom out past 20%', fakeAsync(() => {
+            component.scale = 0.2;
+            tick();
+
+            component.zoomOut();
+            component.zoomOut();
+            component.zoomOut();
+
+            expect(component.scale).toBe(0.2);
+        }));
+
+        it('should show rotate button if not in read only mode', () => {
+            component.readOnly = false;
+            fixture.detectChanges();
+            const rotateButtonElement = element.querySelector('#viewer-rotate-button');
+            expect(rotateButtonElement).not.toEqual(null);
+        });
+
+        it('should not show rotate button by default', () => {
+            const rotateButtonElement = element.querySelector('#viewer-rotate-button');
+            expect(rotateButtonElement).toEqual(null);
+        });
+
+        it('should rotate image by -90 degrees on button click', fakeAsync(() => {
+            component.readOnly = false;
+            spyOn(component, 'rotateImage').and.callThrough();
+            spyOn(component.cropper, 'rotate');
+            fixture.detectChanges();
+            const rotateButtonElement = fixture.debugElement.query(By.css('#viewer-rotate-button'));
+            rotateButtonElement.triggerEventHandler('click', null);
+            tick();
+            expect(component.rotateImage).toHaveBeenCalled();
+            component.rotateImage();
+            expect(component.cropper.rotate).toHaveBeenCalledWith(-90);
+        }));
+
+        it('should display or not the second toolbar if in read only mode', fakeAsync(() => {
+            component.isEditing = true;
+            component.readOnly = true;
+            fixture.detectChanges();
+            let secondaryToolbar = document.querySelector('.adf-secondary-toolbar');
+            expect(secondaryToolbar).toEqual(null);
+
+            component.readOnly = false;
+            fixture.detectChanges();
+            secondaryToolbar = document.querySelector('.adf-secondary-toolbar');
+            expect(secondaryToolbar).not.toEqual(null);
+        }));
+
+        it('should display second toolbar in rotate mode', fakeAsync(() => {
+            component.readOnly = false;
+            expect(component.isEditing).toEqual(false);
+            component.rotateImage();
+            expect(component.isEditing).toEqual(true);
+            fixture.detectChanges();
+            const secondaryToolbar = document.querySelector('.adf-secondary-toolbar');
+            const resetButton = document.querySelector('#viewer-cancel-button');
+            const saveButton = document.querySelector('#viewer-save-button');
+            expect(secondaryToolbar).not.toEqual(null);
+            expect(resetButton).not.toEqual(null);
+            expect(saveButton).not.toEqual(null);
```

Comment:
The steps 1-4 are checking that if you rotate the image the component is going to be in editing mode, so the test is checking 2 things.
I suggest the following:
If that is a precondition that is needed to the rest of the test then instead of the 4 first steps just make your component be in editing mode by doing `component.isEditing = true;`
The first 4 steps can be a different test (should be in editing mode when the image gets rotated)

Reply:
@arditdomi ... thanks for the feedback. @ursedaniel is PTO until next Wednesday.

Could this PR be merged as-is.? If, so, I'm happy to also log in JIRA your suggested unit test improvements so that we don't forget to review again, eg. mid next week (or worst case during code freeze week).

cc @KikoUK @PunchedChimera

Reply:
Hi @montgolfiere I would say it's fine but I don't think we can merge it yet, even without fixing my recommendations as there are e2e in the viewer failing https://travis-ci.com/github/Alfresco/alfresco-ng2-components/jobs/502015163#L1337 
If @KikoUK and @PunchedChimera  can fix the failing e2e, I can take care of the comments made by me

Reply:
This PR cannot be merged for now cos the introduced changes break existing e2es.

Reply:
If we manage to fix the e2e I am happy to fix my comments, it should take me 5 mins to do so. It is just to leave spaces between block of expectations, and separate 2 unit tests into 4 so we achieve that every test is testing one thing and not more. Thanks all

Reply:
Thanks @arditdomi and @KikoUK. 

I just wanted to make sure this PR does not get stuck for 3+ days.

@KikoUK Are you ok to fix e2e impact ?

@arditdomi Thanks for offering to fix the formatting & splitting (as per your suggestions).

Reply:
Great. I notice the "Cloud" e2e still seem to red. First one shows:
```
login SSO ok devopsUsername
ERROR: App subprocessapp down or inaccessible  with status Pending
ERROR: App candidatebaseapp down or inaccessible  with status Pending
ERROR: App simpleapp down or inaccessible  with status Pending
```
I'll restart the first failed job and see what happens. I'm not sure the correct way to get these green and PR merged.

Reply:
I've restarted "Process Cloud e2e" jobs on both PR branch and develop. Now all green.

Hence, will merge.